### PR TITLE
Mutable memory

### PR DESCRIPTION
**Pull request by arcz**
## Description
Closes https://github.com/ethereum/hevm/issues/292.

## Checklist

- [ ] tested locally
- [ ] added automated tests
- [ ] updated the docs
- [ ] updated the changelog

### CODE REVIEW

**Comment by msooseth** [comment]
BTW, the previous failure, `:main --quickcheck-replay=116473 -p "buffer-simplification"` was reproducible only on buggy Z3. I am thinking about going back to old Z3...

**Comment by arcz** [comment]
yeah, the tests now fail quite often and I believe this is all caused by the new z3

File: src/EVM/Types.hs [hunk 1]
```diff
@@ -680,6 +683,14 @@ data FrameState = FrameState
   }
   deriving (Show, Generic)
 
+data Memory s
+  = ConcreteMemory !(STVector s Word8)
+  | SymbolicMemory !(Expr Buf)
+
+instance Show (Memory s) where
+  show (ConcreteMemory _) = "<can't show mutable memory>"
```
Comment (d-xo) [inline src/EVM/Types.hs:691]: why not?

Comment (arcz) [reply src/EVM/Types.hs:691]: because we need to be in ST or IO monad to do that

File: src/EVM.hs [hunk 2]
```diff
@@ -543,7 +548,10 @@ exec1 = do
               burn g_verylow $
                 accessMemoryWord x $ do
                   next
-                  assign (#state % #stack) (readWord (Lit x) mem : xs)
+                  xd <- readMemory (Lit x) (Lit 32)
+                  let w = Expr.readWordFromBytes (Lit 0) xd
+                  assign (#state % #stack) (w : xs)
+                  -- assign (#state % #stack) (readWord (Lit x) mem : xs)
```
Comment (d-xo) [inline src/EVM.hs:554]:
```suggestion
```

File: src/EVM.hs [hunk 3]
```diff
@@ -1932,33 +1967,76 @@ accessMemoryRange f l continue =
...
+            expandMemory (fromIntegral (memOffset' + size'))
+            gets (.state.memory) >>= \case
+              ConcreteMemory mem' -> do
+                let src =
+                      if srcOffset' >= fromIntegral (BS.length b) then
+                        BS.replicate (fromIntegral size') 0
+                      else
+                        BS.take (fromIntegral size') $
+                        padRight (fromIntegral size') $
+                        BS.drop (fromIntegral srcOffset') b
```
Comment (d-xo) [inline src/EVM.hs:1991]: can we use `tryFrom` / `into` here?

File: src/EVM.hs [hunk 4]
```diff
@@ -1932,33 +1967,76 @@ accessMemoryRange f l continue =
...
+            let pastEnd :: Int = (fromIntegral offset + fromIntegral size) - fromIntegral memSize
+            let fromMemSize = if pastEnd > 0 then fromIntegral size - pastEnd else fromIntegral size
 
-readMemory :: Expr EWord -> Expr EWord -> VM -> Expr Buf
-readMemory offset size vm = copySlice offset (Lit 0) size vm.state.memory mempty
+            a <- VUnboxed.freeze $ VUnboxed.Mutable.slice (fromIntegral offset) fromMemSize mem
+            let dataFromMem = BS.pack $ VUnboxed.toList a
+            pure $ ConcreteBuf $ dataFromMem <> BS.replicate (fromIntegral pastEnd) 0
+        _ ->
+          error "readMemory transition"
```
Comment (d-xo) [inline src/EVM.hs:2032]: I think we can freeze memory and call `copySlice` on it here?

**Review by d-xo — COMMENTED** [review]
Aside from the unfortunate proliferation of type parameters this is surprisingly clean. I had expected integrating St to be more involved tbh. Really nice work <3. 

I wonder if it's possible to add some type aliases or newtypes or something to tidy up some of the clutter?

**Comment by d-xo** [comment]
What's left to do here? I have a bunch of larger changes I wanna make and I'd like to get this in first to avoid more merge work for you @arcz

**Review by d-xo — APPROVED** [review]
I think this looks great. No comments. IMO we can merge as soon as CI passes here.

**Pull request merged**